### ISSUE TITLE
Drop `g_guidCheck` and enforce valid GUID

### ISSUE THREAD

**Issue by rmarquis** [etlegacy/etlegacy#1813]
In #1346, the GUID validity check has been moved from the server to the Legacy mod side. The existing `sv_guidCheck` cvar has also been moved to `g_guidCheck` (e4035d4a5c6c481ea91267a2579f6ec8653f36e9).

However, disabling this cvar means clients without any GUID can connect, and such clients include 2.60/2.60b clients without PB enabled. This has several drawbacks, including administration issues (ban acting on GUID) and various features that won't work correctly. As such, allowing invalid GUID is a bad idea, and ideally this cvar should not even exist but be enforced by default - especially on public servers.

The reason some players still use the vanilla client seem to be nothing more than inertia (see discussion in #1346), so some effort is required to incentive them to move to ET:L client before enforcing the GUID check.

The dropping messages have been already extended in e4035d4a5c6c481ea91267a2579f6ec8653f36e9 to give some hints.

A few random ideas for clients without valid GUID:
* print some "Download ET:L client" *always* on screen, that can't be hidden
* allow clients without GUID to connect, but disallow them to join a team
* forced redirect to ETL website (is that even possible?)
* better explanation of why the vanilla client is obsolete, and why ETL client is the way to go
* ...

The 2.60b clients that have PB enabled aren't directly concerned here, but we might incentive them to move to the ET:L client in similar way (without dropping them forcibly).

**Comment by isRyven** [comment]
Week ago or so I hesitated to post my analysis in linked discussion regarding the matter, but i'll post the tldr of the idea now:

As alternative solution to provided action points, you could also implement GUID generation in the mod-land, patching it with legacy specifics, like all other (popular) mods do already anyways. Having control over own GUID implementation, will give such benefits as: 
* Per server GUID - prevent potential GUID hijacking by other servers.
* Server side (qagame) can be responsible for GUID issuing - user identification, sort of registration process.
* Own GUID format (could be even UUID based) - file can be encrypted for safety and forgery prevention.
* Engine-agnostics - no need to strictly rely on engine to provide the GUID anymore.
* Don't break existing mods that happen to be hosted by the ETLegacy, which already provide own GUID implementation and originally able to serve the other clients (granted ETLegacy does not enforce GUID check in the engine anymore).

The motivation - at the moment ETLegacy still issues the GUID on the engine-side, but there is no code anymore to validate the format it issued, it's now the mod's responsibility to validate it - a discrepancy in behavior.

**Comment by rmarquis** [comment]
Yes, I also quickly thought about that possibility, but came to realize the effort required to allow clients without GUID in a safe manner isn't worth it (or even counter productive I'd say). After all, the root of the issue is that some players still use vanilla client for dubious reasons, and that's what we should rather aim to "fix" imho.

The only reason to keep using vanilla client I can think of is playing etpro, but well, I think it's fair to say that reason is less and less valid as time go by, and that any vanilla player that switches to ETL will be a happier player.

**Comment by ensiform** [comment]
@isRyven mod side guid generation has one very huge downside, in that it is not guaranteed to be in the connection userinfo packet.

**Comment by isRyven** [comment]
@ensiform you don't need to store (custom) GUID in userinfo to make identity system to work between mod counterparts. That's the best part of it - the way you implement the authentication is up to you as a mod creator. Like if you're worried about the validation upon first connection, then I doubt that's a very big issue, as you can postpone the authentication to next frames and work out own protocol (eg. etjump).

I am not entirely sure i've understood the @rmarquis point or if he understands mine. My point was to stop covering all mod cases, and just focus on legacy itself. I don't suggest to drop the etkey generation in general though, because you still need to keep compatibility with some mods that might rely on it. But i suggest to handle (custom) guid generation and validation directly in the legacy mod, like other mods do, this will help to ensure that every client playing legacy will always have the (custom) guid there.

**Comment by rmarquis** [comment]
I think I've understood your point, I'm also having a strictly Legacy centered viewpoint here. But rather than looking for answering "How to allow client without GUID by getting out of our ways with mod side GUID generation?", I'm more focused on asking "Why allowing such clients?" in the first place.

To me, giving players the options of either
* allowing vanilla ET client but on the condition that PB is enabled, or
* using the much better ET:Legacy client instead

is a perfectly reasonable solution.

And somehow, incentivizing such players to keep using vanilla ET client in any form (without PB, or even with PB) is counter productive: there is no reason to keep using the old client other than seemingly dubious ones. *But* we might have some issue in making clear why the ET:L client is the better choice on modern machine, but that's a more "marketing" than technical reason here.

**Comment by isRyven** [comment]
I understand. While i'd also force people to switch to ETLegacy for their own sake (as VET experience is awful), i still have came up with the next points:

1. I think using GUID as a mean of a client discrimination would be a wrong direction. Personally, I'd rather teach people about ETLegacy (or alternative), than strictly enforce them using it to just play **legacy** mod. This can be achieved in different ways - as mentioned, through marketing or in-game popups. 
2. If mod can support other clients but ETLegacy, and doesn't depend on internal features, then this exclusiveness doesn't make much sense, especially taking into account both ETLegacy and VET clients are prone to hacks (having pb here or not, make no difference). 
3. You have deliberately mentioned that GUID is only necessary to make the internal sub systems working - making clear it's not about VET clients but the lack of etkey - which, as mentioned, could be solved on mod side, once and for all - in a beneficial way.
4. "Why allowing such clients?" - simple, to provide people the service, there is absolutely no good reason not to. As a mod creator it's in your interest to attract as much as possible people to play your creation through which you can enlighten people about ETLegacy - which have been successfully done with the help of TM for the last several years.
5. If you want to ban VET clients, better do this other way, based on some signature. Why? It's not a written rule that other engines must be responsible for generating etkey - it was always a responsibility of PB to this. Hence, if one would work on another engine alternative, you can't enforce the rule to make someone implement etkey generator, as there is no specification or documentation how to do that compatibly. And no one should really consider ETLegacy as a single source of truth here. If we want to enforce ETKey, there should be a specification for this published. Until then we should not assume there is only two engines exist in the world, where legacy should be played by only one of them.

_ps. There should be a code of conduct, that would forbid any hate against certain communities and minorities. I, and pretty sure other contributors as well, are tired hearing humiliation towards etpro players and VET users, we should put this kind of speech to an end! ETLegacy is first of all community and then engine and mod. It should welcome everyone, for the sake of ET prosperity._

**Comment by rmarquis** [comment]
@isRyven Oh well, what the hell is this? This is not about "banning VET client", and implying it the way you do it here seems really disingenuous. You'll really excuse me if I find that mentioning the "code of conduct" (which is about actual hate and racism) in context of the VET client is.... astonishingly stupid, to say it simply. Seriously.

Honestly, I'd never imagined that requiring (a subset of) a minority of clients to simply enable PB to get a GUID or downloading ETL client would be that controversial, and compared to *actual* racism and discrimination. Especially when there is no actual reason to use that old client.

I'm dumbfounded.

**Comment by isRyven** [comment]
@rmarquis The last part was a joke, don't take it seriously, I am sorry if I have hurt you personally.

And yes I am stupid (should be removed from the project as such) and not afraid to admit it, but I still stand for other things I wrote. As you said:

> I'm more focused on asking "Why allowing such clients?" in the first place.

From my point of view, dictating which client people should use to play the mod shouldn't be the motive here, it's their choice, etkey-less or not. (As I mentioned etkey should not be a mechanism to rely on in the first place.) If one can provide a good and working alternative to VET - people will switch willingly. But would creating a technical limitation for VET clients worth it? Personally, I don't think so (and have stated why).

What my ignorance was trying to say here, is that it's out of scope of this project to fight against vanilla client, the strength should be put into creating a good alternative (one of) and not into building some ideology here. But these are my personal views. But if I just incorrectly (because of my stupidity) seeing the aim of the project, then I will take my words back.

**Comment by rmarquis** [comment]
Closing. `g_guidCheck` is there to stay.
